pep-0712.rst

```diff
@@ -0,0 +1,326 @@
+PEP: 712
+Title: Adding a "converter" parameter to dataclasses.field
+Author: Joshua Cannon <joshdcannon@gmail.com>
+Sponsor: Eric V. Smith <eric at trueblade.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 01-Jan-2023
+Python-Version: 3.13
+Post-History: `27-Dec-2022 <https://mail.python.org/archives/list/typing-sig@python.org/thread/NWZQIINJQZDOCZGO6TGCUP2PNW4PEKNY/>`__,
+              `19-Jan-2023 <https://discuss.python.org/t/add-converter-to-dataclass-field/22956>`__,
+
+Abstract
+========
+
+:pep:`557` added :mod:`dataclasses` to the Python stdlib. :pep:`681` added
+:func:`~py3.11:typing.dataclass_transform` to help type checkers understand
+several common dataclass-like libraries, such as attrs, Pydantic, and object
+relational mapper (ORM) packages such as SQLAlchemy and Django.
+
+A common feature these libraries provide over the standard library
+implementation is the ability for the library to convert arguments given at
+initialization time into the types expected for each field using a
+user-provided conversion function.
+
+Therefore, this PEP adds a ``converter`` parameter to :func:`dataclasses.field`
+(along with the requisite changes to :class:`dataclasses.Field` and
+:func:`~py3.11:typing.dataclass_transform`) to specify the function to use to
+convert the input value for each field to the representation to be stored in
+the dataclass.
+
+Motivation
+==========
+
+There is no existing, standard way for :mod:`dataclasses` or third-party
+dataclass-like libraries to support argument conversion in a type-checkable
+way. To work around this limitation, library authors/users are forced to choose
+to:
+
+* Opt-in to a custom Mypy plugin. These plugins help Mypy understand the
+  conversion semantics, but not other tools.
+* Shift conversion responsibility onto the caller of the dataclass
+  constructor. This can make constructing certain dataclasses unnecessarily
+  verbose and repetitive.
+* Provide a custom ``__init__`` and which declares "wider" parameter types and
+  converts them when setting the appropriate attribute. This not only duplicates
+  the typing annotations between the converter and ``__init__``, but also opts
+  the user out of many of the features :mod:`dataclasses` provides.
+* Not rely on, or ignore type-checking.
```

Comment:
This doesn't make sense to me, as converters would seem to be a runtime feature that dataclasses simply doesn't support. It's not just about typing.

Reply:
I'm having trouble mentally parsing your comment mapped back onto the code (it doesn't help that GitHub has displayed several lines as context so I'm not sure exactly the spot you're commenting on).

Can you be more specific or put it another way?

Reply:
I would tell you but I'm about to board a flight to PyCon :). I'll respond when I'm in SLC tonight. If you're also at PyCon, happy to talk in person too.

Reply:
I'll be there!
How are people finding each other at these things?

Reply:
Never mind, got more time before my flight.

Reading this again, I think my confusion is about the fact that the PEP basically has two goals:

- Adding converter support to `dataclasses` in the stdlib
- Adding converter support to the PEP 681 `dataclass_transform` mechanism

I would expect the PEP to primarily be motivated by the first goal, since that's what the PEP title suggests. And if the PEP is about adding a feature to `dataclasses`, the list entry I commented on doesn't make a lot of sense.

But a lot of your PEP reads like it's primarily motivated by the second goal of extending `dataclass_transform`. That's fine, but it means that there is little motivation for adding this new feature to `dataclasses` in the stdlib. So I think my feedback is that you need to motivate better why we couldn't just add this feature to PEP 681 without changing `dataclasses`.

Reply:
As for PyCon, if you're able I'd encourage you to attend the typing summit (Thursday afternoon) and/or the sprints (Monday onwards); I'll be at both. During the conference itself, people often just meet up at whatever time is convenient. You could perhaps submit a lightning talk or create an open space to talk about your upcoming PEP, so you can get feedback and ideas from more people.

Reply:
Ok I've certainly led you astray then. My motivation is very much for data classes, as I'm not even using dataclass_transform.

Additionally, whether the feature goes in either, type checkers are responsible for checking the values, which is why a large part of the PEP is dedicated to typing.